modules/pam_filter/pam_filter.8
.\"     Title: pam_filter
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.70.1 <http://docbook.sf.net/>
.\"      Date: 06/09/2006
.\"    Manual: Linux\-PAM Manual
.\"    Source: Linux\-PAM Manual
.\"
.TH "PAM_FILTER" "8" "06/09/2006" "Linux\-PAM Manual" "Linux\-PAM Manual"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
pam_filter \- PAM filter module
.SH "SYNOPSIS"
.HP 14
\fBpam_filter.so\fR [debug] [new_term] [non_term] run1|run2 \fIfilter\fR [\fI...\fR]
.SH "DESCRIPTION"
.PP
This module is intended to be a platform for providing access to all of the input/output that passes between the user and the application. It is only suitable for tty\-based and (stdin/stdout) applications.
.PP
To function this module requires
\fIfilters\fR
to be installed on the system. The single filter provided with the module simply transposes upper and lower case letters in the input and output streams. (This can be very annoying and is not kind to termcap based editors).
.PP
Each component of the module has the potential to invoke the desired filter. The filter is always
\fBexecv\fR(2)
with the privilege of the calling application and
\fInot\fR
that of the user. For this reason it cannot usually be killed by the user without closing their session.
.SH "OPTIONS"
.PP
.TP 3n
\fBdebug\fR
Print debug information.
.TP 3n
\fBnew_term\fR
The default action of the filter is to set the
\fIPAM_TTY\fR
item to indicate the terminal that the user is using to connect to the application. This argument indicates that the filter should set
\fIPAM_TTY\fR
to the filtered pseudo\-terminal.
.TP 3n
\fBnon_term\fR
don't try to set the
\fIPAM_TTY\fR
item.
.TP 3n
\fBrunX\fR
In order that the module can invoke a filter it should know when to invoke it. This argument is required to tell the filter when to do this.
.sp
Permitted values for
\fIX\fR
are
\fI1\fR
and
\fI2\fR. These indicate the precise time that the filter is to be run. To understand this concept it will be useful to have read the
\fBpam\fR(3)
manual page. Basically, for each management group there are up to two ways of calling the module's functions. In the case of the
\fIauthentication\fR
and
\fIsession\fR
components there are actually two separate functions. For the case of authentication, these functions are
\fBpam_authenticate\fR(3)
and
\fBpam_setcred\fR(3), here
\fBrun1\fR
means run the filter from the
\fBpam_authenticate\fR
function and
\fBrun2\fR
means run the filter from
\fBpam_setcred\fR. In the case of the session modules,
\fIrun1\fR
implies that the filter is invoked at the
\fBpam_open_session\fR(3)
stage, and
\fIrun2\fR
for
\fBpam_close_session\fR(3).
.sp
For the case of the account component. Either
\fIrun1\fR
or
\fIrun2\fR
may be used.
.sp
For the case of the password component,
\fIrun1\fR
is used to indicate that the filter is run on the first occasion of
\fBpam_chauthtok\fR(3)
(the
\fIPAM_PRELIM_CHECK\fR
phase) and
\fIrun2\fR
is used to indicate that the filter is run on the second occasion (the
\fIPAM_UPDATE_AUTHTOK\fR
phase).
.TP 3n
\fBfilter\fR
The full pathname of the filter to be run and any command line arguments that the filter might expect.
.SH "MODULE SERVICES PROVIDED"
.PP
The services
\fBauth\fR,
\fBaccount\fR,
\fBpassword\fR
and
\fBsession\fR
are supported.
.SH "RETURN VALUES"
.PP
.TP 3n
PAM_SUCCESS
The new filter was set successfull.
.TP 3n
PAM_ABORT
Critical error, immediate abort.
.SH "EXAMPLES"
.PP
Add the following line to
\fI/etc/pam.d/login\fR
to see how to configure login to transpose upper and lower case letters once the user has logged in:
.sp
.RS 3n
.nf
        session required pam_filter.so run1 /lib/security/pam_filter/upperLOWER
      
.fi
.RE
.sp
.SH "SEE ALSO"
.PP

\fBpam.conf\fR(5),
\fBpam.d\fR(8),
\fBpam\fR(8)
.SH "AUTHOR"
.PP
pam_filter was written by Andrew G. Morgan <morgan@kernel.org>.


NEWS
Linux-PAM NEWS -- history of user-visible changes.


* pam_tally: Fix support for large UIDs
* Fixed all problems found by Coverity
* Add manual page for pam_mkhomedir, pam_umask, pam_filter,
  pam_issue, pam_ftp, pam_group, pam_lastlog, pam_listfile,
  pam_localuser, pam_mail, pam_motd, pam_nologin, pam_permit,
  pam_rootok, pam_securetty, pam_shells, pam_userdb, pam_warn,
  pam_time
* The libpam memory debug code was removed

Release 0.99.4.0

* Add test suite
* Fix building of static variants of libpam, libpamc and libpam_misc
* pam_listfile: Add support for password and session management
* pam_exec: New PAM module to execute arbitary commands
* Fix building of a static libpam including all PAM modules
* New/updated translations for: nl, pt, pl, fi, km, tr, uk, fr
* pam_access: Add network(address) / netmask and IPv6 support
* Add manual pages for pam_cracklib, pam_deny and pam_access
* pam_pwdb: This deprecated module was removed
* Manual pages: Major rewrite/cleanup


Release 0.99.3.0

* Fix NULL pointer checks in libpam.so
* pam_succeed_if, pam_group, pam_time: Support netgroup matching
* New translations for: nb, hu, fi, de, es, fr, it, ja, pt_BR, zh_CN, zh_TW
* Audit PAM calls if Linux Audit is available
* Compile upperLOWER and unix_chkpwd as PIE binaries


Release 0.99.2.1

* Fix install of PS, PDF, TXT and HTML files
* pam_mail: Update README
* Use %m consistent
* pam_modutil_getlogin: Fix parsing of PAM_TTY variable


Release 0.99.2.0

* Fix parsing of full path tty name in various modules
* pam_xauth: Look for xauth executable in multiple places
* pam_unix: Disable user check in unix_chkpwd only if real uid
  is 0 (CVE-2005-2977). Log failed password check attempt.
* pam_env: Support /etc/environment again, but don't treat it as
  error if it is missing.
* pam_userdb: Fix memory leak.


Release 0.99.1.0

* Use autoconf/automake/libtool
* Add gettext support
* Add translations for cs, de, es, fr, hu, it, ja, nb, pa, pt_BR,
  pt, zh_CN and zh_TW
* libpam: Remove pam_authenticate_secondary stub
* libpam: Add pam_prompt,pam_vprompt,pam_error,pam_verror,pam_info
  and pam_vinfo functions for use by modules as extension
* libpam: Add pam_syslog function for unified syslog messages from
  PAM modules
* libpam: Moved functions from pammodutil to libpam
* pam_umask: New module for setting umask from GECOS field, /etc/login.defs
  or /etc/default/login
* pam_echo: New PAM module for message output
* pam_userdb: Fix regression (crash when crypt param not specified)
* pam_limits: Fix regression from RLIMIT_NICE support (wrong limit
  values for other limits are applied)
* pam_access: Support for NULL tty - matches ALL and NONE keywords
* pam_lastlog: Enable log to wtmp by default.  Add "nowtmp" option
* pam_radius: This module was removed


modules/pam_filter/pam_filter.c
/*
 * $Id$
 *
 * written by Andrew Morgan <morgan@transmeta.com> with much help from
 * Richard Stevens' UNIX Network Programming book.
 */

#include "config.h"

#include <stdlib.h>
#include <syslog.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>

#include <stdio.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <termios.h>

#include <signal.h>

#define PAM_SM_AUTH
#define PAM_SM_ACCOUNT
#define PAM_SM_SESSION
#define PAM_SM_PASSWORD

#include <security/pam_modules.h>
#include <security/pam_ext.h>
#include "pam_filter.h"

/* ------ some tokens used for convenience throughout this file ------- */

#define FILTER_DEBUG     01
#define FILTER_RUN1      02
#define FILTER_RUN2      04
#define NEW_TERM        010
#define NON_TERM        020

/* -------------------------------------------------------------------- */

/* log errors */

#include <stdarg.h>

#define TERMINAL_LEN 12

static int
master (const pam_handle_t *pamh, char *terminal)
/*
 * try to open all of the terminals in sequence return first free one,
 * or -1
 */
{
     const char ptys[] = "pqrs", *pty = ptys;
     const char hexs[] = "0123456789abcdef", *hex;
     struct stat tstat;
     int fd;

     strcpy(terminal, "/dev/pty??");

     while (*pty) {                   /* step through four types */
	  terminal[8] = *pty++;
	  terminal[9] = '0';
	  if (stat(terminal,&tstat) < 0) {
	       pam_syslog(pamh, LOG_WARNING,
			  "unknown pseudo terminal: %s", terminal);
	       break;
	  }
	  for (hex = hexs; *hex; ) {  /* step through 16 of these */
	       terminal[9] = *hex++;
	       if ((fd = open(terminal, O_RDWR)) >= 0) {
		    return fd;
	       }
	  }
     }

     /* no terminal found */

     return -1;
}

static int process_args(pam_handle_t *pamh
			, int argc, const char **argv, const char *type
			, char ***evp, const char **filtername)
{
    int ctrl=0;

    while (argc-- > 0) {
	if (strcmp("debug",*argv) == 0) {
	    ctrl |= FILTER_DEBUG;
	} else if (strcmp("new_term",*argv) == 0) {
	    ctrl |= NEW_TERM;
	} else if (strcmp("non_term",*argv) == 0) {
	    ctrl |= NON_TERM;
	} else if (strcmp("run1",*argv) == 0) {
	    ctrl |= FILTER_RUN1;
	    if (argc <= 0) {
		pam_syslog(pamh, LOG_ALERT, "no run filter supplied");
	    } else
		break;
	} else if (strcmp("run2",*argv) == 0) {
	    ctrl |= FILTER_RUN2;
	    if (argc <= 0) {
		pam_syslog(pamh, LOG_ALERT, "no run filter supplied");
	    } else
		break;
	} else {
	    pam_syslog(pamh, LOG_ERR, "unrecognized option: %s", *argv);
	}
	++argv;                   /* step along list */
    }

    if (argc < 0) {
	/* there was no reference to a filter */
	*filtername = NULL;
	*evp = NULL;
    } else {
	char **levp;
	const char *user = NULL;
	const void *tmp;
	int i,size, retval;

	*filtername = *++argv;
	if (ctrl & FILTER_DEBUG) {
	    pam_syslog(pamh, LOG_DEBUG, "will run filter %s", *filtername);
	}

	levp = (char **) malloc(5*sizeof(char *));
	if (levp == NULL) {
	    pam_syslog(pamh, LOG_CRIT, "no memory for environment of filter");
	    return -1;
	}

	for (size=i=0; i<argc; ++i) {
	    size += strlen(argv[i])+1;
	}

	/* the "ARGS" variable */

#define ARGS_OFFSET    5                          /*  strlen('ARGS=');  */
#define ARGS_NAME      "ARGS="

	size += ARGS_OFFSET;

	levp[0] = (char *) malloc(size);
	if (levp[0] == NULL) {
	    pam_syslog(pamh, LOG_CRIT, "no memory for filter arguments");
	    if (levp) {
		free(levp);
	    }
	    return -1;
	}

	strncpy(levp[0],ARGS_NAME,ARGS_OFFSET);
	for (i=0,size=ARGS_OFFSET; i<argc; ++i) {
	    strcpy(levp[0]+size, argv[i]);
	    size += strlen(argv[i]);
	    levp[0][size++] = ' ';
	}
	levp[0][--size] = '\0';                    /* <NUL> terminate */

	/* the "SERVICE" variable */

#define SERVICE_OFFSET    8                    /*  strlen('SERVICE=');  */
#define SERVICE_NAME      "SERVICE="

	retval = pam_get_item(pamh, PAM_SERVICE, &tmp);
	if (retval != PAM_SUCCESS || tmp == NULL) {
	    pam_syslog(pamh, LOG_CRIT, "service name not found");
	    if (levp) {
		free(levp[0]);
		free(levp);
	    }
	    return -1;
	}
	size = SERVICE_OFFSET+strlen(tmp);

	levp[1] = (char *) malloc(size+1);
	if (levp[1] == NULL) {
	    pam_syslog(pamh, LOG_CRIT, "no memory for service name");
	    if (levp) {
		free(levp[0]);
		free(levp);
	    }
	    return -1;
	}

	strncpy(levp[1],SERVICE_NAME,SERVICE_OFFSET);
	strcpy(levp[1]+SERVICE_OFFSET, tmp);
	levp[1][size] = '\0';                      /* <NUL> terminate */

	/* the "USER" variable */

#define USER_OFFSET    5                          /*  strlen('USER=');  */
#define USER_NAME      "USER="

	pam_get_user(pamh, &user, NULL);
	if (user == NULL) {
	    user = "<unknown>";
	}
	size = USER_OFFSET+strlen(user);

	levp[2] = (char *) malloc(size+1);
	if (levp[2] == NULL) {
	    pam_syslog(pamh, LOG_CRIT, "no memory for user's name");
	    if (levp) {
		free(levp[1]);
		free(levp[0]);
		free(levp);
	    }
	    return -1;
	}

	strncpy(levp[2],USER_NAME,USER_OFFSET);
	strcpy(levp[2]+USER_OFFSET, user);
	levp[2][size] = '\0';                      /* <NUL> terminate */

	/* the "USER" variable */

#define TYPE_OFFSET    5                          /*  strlen('TYPE=');  */
#define TYPE_NAME      "TYPE="

	size = TYPE_OFFSET+strlen(type);

	levp[3] = (char *) malloc(size+1);
	if (levp[3] == NULL) {
	    pam_syslog(pamh, LOG_CRIT, "no memory for type");
	    if (levp) {
		free(levp[2]);
		free(levp[1]);
		free(levp[0]);
		free(levp);
	    }
	    return -1;
	}

	strncpy(levp[3],TYPE_NAME,TYPE_OFFSET);
	strcpy(levp[3]+TYPE_OFFSET, type);
	levp[3][size] = '\0';                      /* <NUL> terminate */

	levp[4] = NULL;	                     /* end list */

	*evp = levp;
    }

    if ((ctrl & FILTER_DEBUG) && *filtername) {
	char **e;

	pam_syslog(pamh, LOG_DEBUG, "filter[%s]: %s", type, *filtername);
	pam_syslog(pamh, LOG_DEBUG, "environment:");
	for (e=*evp; e && *e; ++e) {
	    pam_syslog(pamh, LOG_DEBUG, "  %s", *e);
	}
    }

    return ctrl;
}

static void free_evp(char *evp[])
{
    int i;

    if (evp)
	for (i=0; i<4; ++i) {
	    if (evp[i])
		free(evp[i]);
	}
    free(evp);
}

static int
set_filter (pam_handle_t *pamh, int flags UNUSED, int ctrl,
	    const char **evp, const char *filtername)
{
    int status=-1;
    char terminal[TERMINAL_LEN];
    struct termios stored_mode;           /* initial terminal mode settings */
    int fd[2], child=0, child2=0, aterminal;

    if (filtername == NULL || *filtername != '/') {
	pam_syslog(pamh, LOG_ALERT,
		   "filtername not permitted; full pathname required");
	return PAM_ABORT;
    }

    if (!isatty(STDIN_FILENO) || !isatty(STDOUT_FILENO)) {
	aterminal = 0;
    } else {
	aterminal = 1;
    }

    if (aterminal) {

	/* open the master pseudo terminal */

	fd[0] = master(pamh,terminal);
	if (fd[0] < 0) {
	    pam_syslog(pamh, LOG_CRIT, "no master terminal");
	    return PAM_AUTH_ERR;
	}

	/* set terminal into raw mode.. remember old mode so that we can
	   revert to it after the child has quit. */

	/* this is termios terminal handling... */

	if ( tcgetattr(STDIN_FILENO, &stored_mode) < 0 ) {
	    pam_syslog(pamh, LOG_CRIT, "couldn't copy terminal mode: %m");
	    /* in trouble, so close down */
	    close(fd[0]);
	    return PAM_ABORT;
	} else {
	    struct termios t_mode = stored_mode;

	    t_mode.c_iflag = 0;            /* no input control */
	    t_mode.c_oflag &= ~OPOST;      /* no ouput post processing */

	    /* no signals, canonical input, echoing, upper/lower output */
#ifdef XCASE
	    t_mode.c_lflag &= ~(XCASE);
#endif
	    t_mode.c_lflag &= ~(ISIG|ICANON|ECHO);
	    t_mode.c_cflag &= ~(CSIZE|PARENB);  /* no parity */
	    t_mode.c_cflag |= CS8;              /* 8 bit chars */

	    t_mode.c_cc[VMIN] = 1; /* number of chars to satisfy a read */
	    t_mode.c_cc[VTIME] = 0;          /* 0/10th second for chars */

	    if ( tcsetattr(STDIN_FILENO, TCSAFLUSH, &t_mode) < 0 ) {
		pam_syslog(pamh, LOG_WARNING,
			   "couldn't put terminal in RAW mode: %m");
		close(fd[0]);
		return PAM_ABORT;
	    }

	    /*
	     * NOTE: Unlike the stream socket case here the child
	     * opens the slave terminal as fd[1] *after* the fork...
	     */
	}
    } else {

	/*
	 * not a terminal line so just open a stream socket fd[0-1]
	 * both set...
	 */

	if ( socketpair(AF_UNIX, SOCK_STREAM, 0, fd) < 0 ) {
	    pam_syslog(pamh, LOG_CRIT, "couldn't open a stream pipe: %m");
	    return PAM_ABORT;
	}
    }

    /* start child process */

    if ( (child = fork()) < 0 ) {

	pam_syslog(pamh, LOG_WARNING, "first fork failed: %m");
	if (aterminal) {
		(void) tcsetattr(STDIN_FILENO, TCSAFLUSH, &stored_mode);
	}

	return PAM_AUTH_ERR;
    }

    if ( child == 0 ) {                  /* child process *is* application */

	if (aterminal) {

	    /* close the controlling tty */

#if defined(__hpux) && defined(O_NOCTTY)
	    int t = open("/dev/tty", O_RDWR|O_NOCTTY);
#else
	    int t = open("/dev/tty",O_RDWR);
	    if (t > 0) {
		(void) ioctl(t, TIOCNOTTY, NULL);
		close(t);
	    }
#endif /* defined(__hpux) && defined(O_NOCTTY) */

	    /* make this process it's own process leader */
	    if (setsid() == -1) {
		pam_syslog(pamh, LOG_WARNING,
			   "child cannot become new session: %m");
		return PAM_ABORT;
	    }

	    /* find slave's name */
	    terminal[5] = 't';             /* want to open slave terminal */
	    fd[1] = open(terminal, O_RDWR);
	    close(fd[0]);      /* process is the child -- uses line fd[1] */

	    if (fd[1] < 0) {
		pam_syslog(pamh, LOG_WARNING,
			   "cannot open slave terminal: %s: %m", terminal);
		return PAM_ABORT;
	    }

	    /* initialize the child's terminal to be the way the
	       parent's was before we set it into RAW mode */

	    if ( tcsetattr(fd[1], TCSANOW, &stored_mode) < 0 ) {
		pam_syslog(pamh, LOG_WARNING,
			   "cannot set slave terminal mode: %s: %m", terminal);
		close(fd[1]);
		return PAM_ABORT;
	    }

	} else {

	    /* nothing to do for a simple stream socket */

	}

	/* re-assign the stdin/out to fd[1] <- (talks to filter). */

	if ( dup2(fd[1],STDIN_FILENO) != STDIN_FILENO ||
	     dup2(fd[1],STDOUT_FILENO) != STDOUT_FILENO ||
	     dup2(fd[1],STDERR_FILENO) != STDERR_FILENO )  {
	    pam_syslog(pamh, LOG_WARNING,
		       "unable to re-assign STDIN/OUT/ERR: %m");
	    close(fd[1]);
	    return PAM_ABORT;
	}

	/* make sure that file descriptors survive 'exec's */

	if ( fcntl(STDIN_FILENO, F_SETFD, 0) ||
	     fcntl(STDOUT_FILENO,F_SETFD, 0) ||
	     fcntl(STDERR_FILENO,F_SETFD, 0) ) {
	    pam_syslog(pamh, LOG_WARNING,
		       "unable to re-assign STDIN/OUT/ERR: %m");
	    return PAM_ABORT;
	}

	/* now the user input is read from the parent/filter: forget fd */

	close(fd[1]);

	/* the current process is now aparently working with filtered
	   stdio/stdout/stderr --- success! */

	return PAM_SUCCESS;
    }

    /*
     * process is the parent here. So we can close the application's
     * input/output
     */

    close(fd[1]);

    /* Clear out passwords... there is a security problem here in
     * that this process never executes pam_end.  Consequently, any
     * other sensitive data in this process is *not* explicitly
     * overwritten, before the process terminates */

    (void) pam_set_item(pamh, PAM_AUTHTOK, NULL);
    (void) pam_set_item(pamh, PAM_OLDAUTHTOK, NULL);

    /* fork a copy of process to run the actual filter executable */

    if ( (child2 = fork()) < 0 ) {

	pam_syslog(pamh, LOG_WARNING, "filter fork failed: %m");
	child2 = 0;

    } else if ( child2 == 0 ) {              /* exec the child filter */

	if ( dup2(fd[0],APPIN_FILENO) != APPIN_FILENO ||
	     dup2(fd[0],APPOUT_FILENO) != APPOUT_FILENO ||
	     dup2(fd[0],APPERR_FILENO) != APPERR_FILENO )  {
	    pam_syslog(pamh, LOG_WARNING,
		       "unable to re-assign APPIN/OUT/ERR: %m");
	    close(fd[0]);
	    exit(1);
	}

	/* make sure that file descriptors survive 'exec's */

	if ( fcntl(APPIN_FILENO, F_SETFD, 0) == -1 ||
	     fcntl(APPOUT_FILENO,F_SETFD, 0) == -1 ||
	     fcntl(APPERR_FILENO,F_SETFD, 0) == -1 ) {
	    pam_syslog(pamh, LOG_WARNING,
		       "unable to retain APPIN/OUT/ERR: %m");
	    close(APPIN_FILENO);
	    close(APPOUT_FILENO);
	    close(APPERR_FILENO);
	    exit(1);
	}

	/* now the user input is read from the parent through filter */

	execle(filtername, "<pam_filter>", NULL, evp);

	/* getting to here is an error */

	pam_syslog(pamh, LOG_ALERT, "filter: %s: %m", filtername);

    } else {           /* wait for either of the two children to exit */

	while (child && child2) {    /* loop if there are two children */
	    int lstatus=0;
	    int chid;

	    chid = wait(&lstatus);
	    if (chid == child) {

		if (WIFEXITED(lstatus)) {            /* exited ? */
		    status = WEXITSTATUS(lstatus);
		} else if (WIFSIGNALED(lstatus)) {   /* killed ? */
		    status = -1;
		} else
		    continue;             /* just stopped etc.. */
		child = 0;        /* the child has exited */

	    } else if (chid == child2) {
		/*
		 * if the filter has exited. Let the child die
		 * naturally below
		 */
		if (WIFEXITED(lstatus) || WIFSIGNALED(lstatus))
		    child2 = 0;
	    } else {

		pam_syslog(pamh, LOG_ALERT,
			   "programming error <chid=%d,lstatus=%x> "
			   "in file %s at line %d",
			   chid, lstatus, __FILE__, __LINE__);
		child = child2 = 0;
		status = -1;

	    }
	}
    }

    close(fd[0]);

    /* if there is something running, wait for it to exit */

    while (child || child2) {
	int lstatus=0;
	int chid;

	chid = wait(&lstatus);

	if (child && chid == child) {

	    if (WIFEXITED(lstatus)) {            /* exited ? */
		status = WEXITSTATUS(lstatus);
	    } else if (WIFSIGNALED(lstatus)) {   /* killed ? */
		status = -1;
	    } else
		continue;             /* just stopped etc.. */
	    child = 0;        /* the child has exited */

	} else if (child2 && chid == child2) {

	    if (WIFEXITED(lstatus) || WIFSIGNALED(lstatus))
		child2 = 0;

	} else {

	    pam_syslog(pamh, LOG_ALERT,
		       "programming error <chid=%d,lstatus=%x> "
		       "in file %s at line %d",
		       chid, lstatus, __FILE__, __LINE__);
	    child = child2 = 0;
	    status = -1;

	}
    }

    if (aterminal) {
	/* reset to initial terminal mode */
	    (void) tcsetattr(STDIN_FILENO, TCSANOW, &stored_mode);
    }

    if (ctrl & FILTER_DEBUG) {
	pam_syslog(pamh, LOG_DEBUG, "parent process exited");      /* clock off */
    }

    /* quit the parent process, returning the child's exit status */

    exit(status);
    return status; /* never reached, to make gcc happy */
}

static int set_the_terminal(pam_handle_t *pamh)
{
    const void *tty;

    if (pam_get_item(pamh, PAM_TTY, &tty) != PAM_SUCCESS
	|| tty == NULL) {
	tty = ttyname(STDIN_FILENO);
	if (tty == NULL) {
	    pam_syslog(pamh, LOG_ERR, "couldn't get the tty name");
	    return PAM_ABORT;
	}
	if (pam_set_item(pamh, PAM_TTY, tty) != PAM_SUCCESS) {
	    pam_syslog(pamh, LOG_ERR, "couldn't set tty name");
	    return PAM_ABORT;
	}
    }
    return PAM_SUCCESS;
}

static int need_a_filter(pam_handle_t *pamh
			 , int flags, int argc, const char **argv
			 , const char *name, int which_run)
{
    int ctrl;
    char **evp;
    const char *filterfile;
    int retval;

    ctrl = process_args(pamh, argc, argv, name, &evp, &filterfile);
    if (ctrl == -1) {
	return PAM_AUTHINFO_UNAVAIL;
    }

    /* set the tty to the old or the new one? */

    if (!(ctrl & NON_TERM) && !(ctrl & NEW_TERM)) {
	retval = set_the_terminal(pamh);
	if (retval != PAM_SUCCESS) {
	    pam_syslog(pamh, LOG_ERR, "tried and failed to set PAM_TTY");
	}
    } else {
	retval = PAM_SUCCESS;  /* nothing to do which is always a success */
    }

    if (retval == PAM_SUCCESS && (ctrl & which_run)) {
	retval = set_filter(pamh, flags, ctrl
			    , (const char **)evp, filterfile);
    }

    if (retval == PAM_SUCCESS
	&& !(ctrl & NON_TERM) && (ctrl & NEW_TERM)) {
	retval = set_the_terminal(pamh);
	if (retval != PAM_SUCCESS) {
	    pam_syslog(pamh, LOG_ERR,
		       "tried and failed to set new terminal as PAM_TTY");
	}
    }

    free_evp(evp);

    if (ctrl & FILTER_DEBUG) {
	pam_syslog(pamh, LOG_DEBUG, "filter/%s, returning %d", name, retval);
	pam_syslog(pamh, LOG_DEBUG, "[%s]", pam_strerror(pamh, retval));
    }

    return retval;
}

/* ----------------- public functions ---------------- */

/*
 * here are the advertised access points ...
 */

/* ------------------ authentication ----------------- */

PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh
				   , int flags, int argc, const char **argv)
{
    return need_a_filter(pamh, flags, argc, argv
			 , "authenticate", FILTER_RUN1);
}

PAM_EXTERN int pam_sm_setcred(pam_handle_t *pamh, int flags
			      , int argc, const char **argv)
{
    return need_a_filter(pamh, flags, argc, argv, "setcred", FILTER_RUN2);
}

/* --------------- account management ---------------- */

PAM_EXTERN int pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc,
                              const char **argv)
{
    return need_a_filter(pamh, flags, argc, argv
			 , "setcred", FILTER_RUN1|FILTER_RUN2 );
}

/* --------------- session management ---------------- */

PAM_EXTERN int pam_sm_open_session(pam_handle_t *pamh, int flags
				   , int argc, const char **argv)
{
    return need_a_filter(pamh, flags, argc, argv
			 , "open_session", FILTER_RUN1);
}

PAM_EXTERN int pam_sm_close_session(pam_handle_t *pamh, int flags
			      , int argc, const char **argv)
{
    return need_a_filter(pamh, flags, argc, argv
			 , "close_session", FILTER_RUN2);
}

/* --------- updating authentication tokens --------- */


PAM_EXTERN int pam_sm_chauthtok(pam_handle_t *pamh, int flags
				, int argc, const char **argv)
{
    int runN;

    if (flags & PAM_PRELIM_CHECK)
	runN = FILTER_RUN1;
    else if (flags & PAM_UPDATE_AUTHTOK)
	runN = FILTER_RUN2;
    else {
	pam_syslog(pamh, LOG_ERR, "unknown flags for chauthtok (0x%X)", flags);
	return PAM_TRY_AGAIN;
    }

    return need_a_filter(pamh, flags, argc, argv, "chauthtok", runN);
}

#ifdef PAM_STATIC

/* ------------ stuff for static modules ------------ */

struct pam_module _pam_filter_modstruct = {
    "pam_filter",
    pam_sm_authenticate,
    pam_sm_setcred,
    pam_sm_acct_mgmt,
    pam_sm_open_session,
    pam_sm_close_session,
    pam_sm_chauthtok,
};

#endif
